### PR TITLE
[RFC] Indent with tabs

### PR DESCRIPTION
**Pull request by rhengles**
I'm really sorry for having to create another PR, don't want to spam this project, but Github doesn't allow me to reopen the previous PR #920 because the branch has been rebased while the PR was closed.

### CODE REVIEW

File: src/doc-printer.js [hunk 1]
```diff
@@ -3,6 +3,33 @@
...
+    align: ind.align
+  };
+}
+
+function align(ind, n) {
+  return {
+    indent: ind.indent,
+    align: {
+      spaces: ind.align.spaces + n,
+      tabs: ind.align.tabs + (n ? 1 : 0)
```
Comment (rhengles) [inline src/doc-printer.js]: @vjeux :

> Question to you, if you have nested ternaries and other indentation, you're going to have `<tab><space><tab><space>`. Isn't it an issue in practice?

When printing code indented with tabs, my current solution replaces all calls to `align()` with exactly one tab - doesn't matter how many `spaces` it was aligned with, unless it was zero, then it prints zero tabs also.

Comment (vjeux) [inline src/doc-printer.js]: What happens for code like

```js
function f() {
  a
    ? function() {
      a
        ? function() {
          a
          }
        : c;
      }
    : c;
}
```

Comment (vjeux) [inline src/doc-printer.js]: I'm especially curious if you have tab width = 3.

The inner one is
```
<indent><indent><2 space><indent><2 space><indent>
```

If you do tabWidth = 3 then you have

```
tab(3) tab(3) space(2) tab(3) space(2) tab(3)
```

if you output it like this, then the tab after the spaces is going to be width = 1 in order to align, which is incorrect.

What you are doing instead is

```
tab(3) tab(3) tab(3) space(2) space(2)
```

which seems to be correct!

**Comment by vjeux** [comment]
@jlongster I actually think that the solution we sketched out isn't going to work well for nested spaces that are "fixed". It seems also like a good thing to have two different apis for fixed and variable indentation.

Could you review this pull request again? Thanks!

**Comment by jlongster** [comment]
I will review this soon, sorry. It's a bigger thing to talk about so after I go through all the small issues/PRs I end up not having time to tackle this :) I will try to look at this later tonight or tomorrow morning.

It sounds like we're converging on a solution and this is looking good. I'm not stuck on the approach I was suggesting before.

**Comment by jlongster** [comment]
I am *so* sorry I neglected this. A few things came up the last few weeks and I had to focus on client or other work. I'll be looking through prettier again over the next few days, and I'll look at this issue later tonight and review it. (It's probably ready, but I'll take one pass at it)

**Comment by rhengles** [comment]
I will push now a rebase on top of the recent PRs. 👍

File: src/doc-printer.js [hunk 2]
```diff
@@ -35,7 +62,11 @@ function fits(next, restCommands, width) {
 
           break;
         case "indent":
-          cmds.push([ind + doc.n, mode, doc.contents]);
+          cmds.push([indent(ind), mode, doc.contents]);
```
Comment (jlongster) [inline src/doc-printer.js]: It's a little confusing to name these functions `indent` and `align`, the same as the "commands". Can you possibly rename these? At first I thought it was calling the command and was very confused. Maybe something like `makeIndent`?

**Comment by jlongster** [comment]
Alright, thanks for describing the reasons for this algorithm and being patient. This looks good to me. I made one small comment where I think there could be some clarification, but otherwise this is fine.

**Comment by rhengles** [comment]
@jlongster Fixed! Thank you very much! 😃 Please forgive me for my exaggerated enthusiasm sometimes... 😄

**Comment by jlongster** [comment]
That was quick! No worries at all, I took it as just being thorough. Let's merge this and see if any problems come up!

**Comment by respectTheCode** [comment]
are there plans to do a release with this?

**Comment by jlongster** [comment]
@respectTheCode Definitely, not sure exactly when though. In the next few days at the latest.

**Pull request merged**